pkg/manifest/manifest.go

```diff
@@ -15,25 +17,60 @@ import (
 	"k8s.io/client-go/kubernetes/scheme"
 )
 
+// ResourceIdYaml is used to parse out the information
+// that uniquely identifies a Kubernetes resource.
+type ResourceIdYaml struct {
+	APIVersion string `yaml:"apiVersion"`
+	Kind       string `yaml:"kind"`
+	Metadata   struct {
+		Name      string `yaml:"name"`
+		Namespace string `yaml:"namespace"`
+	}
+}
+
+// ResourceId uniquely identifies a Kubernetes resource.
+// It is used to identify any duplicate resources within
+// a given set of manifests.
+type ResourceId struct {
+	Group     string
+	Kind      string
+	Name      string
+	Namespace string
+}
+
 // Manifest stores Kubernetes object in Raw from a file.
-// It stores the GroupVersionKind for the manifest.
-// Raw and Obj should always be kept in sync such that
-// each provides the same data but in different formats.
-// To ensure Raw and Obj are always in sync, they should not
-// be set directly but rather only be set by calling
-// either method ManifestsFromFiles or ParseManifests.
+// It stores the ResourceId and the GroupVersionKind for
+// the manifest. Raw and Obj should always be kept in sync
+// such that each provides the same data but in different
+// formats. To ensure Raw and Obj are always in sync, they
+// should not be set directly but rather only be set by
+// calling either method ManifestsFromFiles or
+// ParseManifests.
 type Manifest struct {
 	// OriginalFilename is set to the filename this manifest was loaded from.
-	// It is not guaranteed to be set or be unique, but we will set it when
-	// loading from disk to provide better debuggability.
+	// It is not guaranteed to be set or be unique, but will be set when
+	// loading from disk to provide a better debug capability.
 	OriginalFilename string
 
+	resourceId ResourceId
+
 	Raw []byte
 	GVK schema.GroupVersionKind
 
 	Obj *unstructured.Unstructured
 }
 
+func (r ResourceId) fullName() string {
+	if len(r.Namespace) == 0 {
+		return r.Name
+	}
+	return r.Namespace + "/" + r.Name
+}
+
+func (r ResourceId) String() string {
+	return fmt.Sprintf("Group: \"%s\" Kind: \"%s\" Name: \"%s\"", r.Group, r.Kind, r.fullName())
```

Comment:
nit: Instead of the `fullName` collapse, I'd rather use:

```go
func (r ResourceId) String() string {
  if len(r.Namespace) == 0 {
    return fmt.Sprintf("Group: %q Kind: %q Name: %q", r.Group, r.Kind, r.Name)
  } else {
    return fmt.Sprintf("Group: %q Kind: %q Namespace: %q Name: %q", r.Group, r.Kind, r.Namespace, r.Name)
  }
}
```

or some such, so readers don't have to mentally split on the slash.  And while I prefer `%q` if we want quotes, I'm not actually sure we need quotes on these, because they're already covered by domain-name restrictions.  So I'm fine with `%q`, and I'm fine with `%s`, but I'd rather avoid `"%s"`.

Reply:
Done